src/radar/radar/toyota_radar_controller.py

```diff
@@ -0,0 +1,124 @@
+import cantools
+import rclpy
+from opencaret_msgs.msg import CanMessage, RadarTrack, RadarTrackAccel, RadarTracks
+from rclpy.node import Node
+from util import util
+import os.path
+import opendbc
+
+class ECU:
+    CAM = 0  # camera
+    DSU = 1  # driving support unit
+    APGS = 2  # advanced parking guidance system
+
+
+class CAR:
+    PRIUS = 0
+    LEXUS_RXH = 1
+    RAV4 = 2
+    RAV4H = 3
+    COROLLA = 4
+
+
+STATIC_MSGS = [
+    (0x141, ECU.DSU, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.RAV4, CAR.COROLLA), 1, 2, '\x00\x00\x00\x46'),
+    (0x128, ECU.DSU, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.RAV4, CAR.COROLLA), 1, 3,
+     '\xf4\x01\x90\x83\x00\x37'),
+    (0x283, ECU.DSU, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.RAV4, CAR.COROLLA), 0, 3,
+     '\x00\x00\x00\x00\x00\x00\x8c'),
+    (0x344, ECU.DSU, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.RAV4, CAR.COROLLA), 0, 5,
+     '\x00\x00\x01\x00\x00\x00\x00\x50'),
+    (0x160, ECU.DSU, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.RAV4, CAR.COROLLA), 1, 7,
+     '\x00\x00\x08\x12\x01\x31\x9c\x51'),
+    (0x161, ECU.DSU, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.RAV4, CAR.COROLLA), 1, 7,
+     '\x00\x1e\x00\x00\x00\x80\x07'),
+    (0x365, ECU.DSU, (CAR.RAV4, CAR.COROLLA), 0, 20, '\x00\x00\x00\x80\xfc\x00\x08'),
+    (0x366, ECU.DSU, (CAR.RAV4, CAR.COROLLA), 0, 20, '\x00\x72\x07\xff\x09\xfe\x00'),
+    (0x4CB, ECU.DSU, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.RAV4, CAR.COROLLA), 0, 100,
+     '\x0c\x00\x00\x00\x00\x00\x00\x00'),
+]
+
+
+class ToyotaRadarController(Node):
+    """
+    This radar controller is hardcoded to work only with the Toyota Corolla/Rav4/Camry 2017 Denso unit
+    """
+
+    def __init__(self):
+        super().__init__('radar')
+        # self.tracks_pub = self.create_publisher(String, 'radar_tracks')
+        self.can_pub = self.create_publisher(CanMessage, 'can_send')
+        self.adas_db = cantools.db.load_file(os.path.join(opendbc.DBC_PATH, 'toyota_prius_2017_adas.dbc'))
+        self.db = cantools.db.load_file(os.path.join(opendbc.DBC_PATH, 'toyota_prius_2017_pt_generated.dbc'))
+
+        self.can_sub = self.create_subscription(CanMessage, 'can_recv', self.on_can_message)
+        self.radar_pub = self.create_publisher(RadarTracks, 'radar_tracks')
+        self.power_on_timer = self.create_timer(1.0/100, self.power_on_radar)
```

Comment:
Is this a persistent timer constantly executing power_on_radar()? If so, why is this necessary?

Reply:
Toyota wants it this way :man_shrugging:

Reply:
Could you add a TODO comment here about how it's not clear why this is needed but that it does continually fire? Definitely should look into why this is at some point... it's sending out 100 messages a second just to keep the radar 'on'.

BTW Timers could be one-shot, but the default mode is perpetual.
http://wiki.ros.org/roscpp/Overview/Timers

Reply:
Feel free to add one and lets merge this PR - need to actually test it.